src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,344 @@
+import org.whispersystems.textsecuregcm.redis.ReplicatedJedisPool;
+import org.whispersystems.textsecuregcm.util.Constants;
+import org.whispersystems.textsecuregcm.util.Hex;
+import org.whispersystems.textsecuregcm.util.Util;
+import redis.clients.jedis.Jedis;
+
+import javax.ws.rs.ProcessingException;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+
+import static com.codahale.metrics.MetricRegistry.name;
+
+public class DirectoryReconciler implements Managed, Runnable {
+
+  private static final Logger logger = LoggerFactory.getLogger(DirectoryReconciler.class);
+
+  private static final MetricRegistry metricRegistry      = SharedMetricRegistries.getOrCreate(Constants.METRICS_NAME);
+  private static final Timer          readChunkTimer      = metricRegistry.timer(name(DirectoryReconciler.class, "readChunk"));
+  private static final Timer          sendChunkTimer      = metricRegistry.timer(name(DirectoryReconciler.class, "sendChunk"));
+  private static final Meter          sendChunkErrorMeter = metricRegistry.meter(name(DirectoryReconciler.class, "sendChunkError"));
+
+  private static final String WORKERS_KEY       = "directory_reconciliation_workers";
+  private static final String ACTIVE_WORKER_KEY = "directory_reconciliation_active_worker";
+  private static final String LAST_NUMBER_KEY   = "directory_reconciliation_last_number";
+  private static final String CACHED_COUNT_KEY  = "directory_reconciliation_cached_count";
+  private static final String ACCELERATE_KEY    = "directory_reconciliation_accelerate";
+
+  private static final long   CACHED_COUNT_TTL_MS       = 21600_000L;
+  private static final long   WORKER_TTL_MS             = 120_000L;
+  private static final long   PERIOD                    = 86400_000L;
+  private static final long   CHUNK_INTERVAL            = 42_000L;
+  private static final long   ACCELERATE_CHUNK_INTERVAL = 500L;
+  private static final double JITTER_MAX                = 0.20;
+
+  private final String              serverApiUrl;
+  private final ReplicatedJedisPool jedisPool;
+  private final AccountsManager     accountsManager;
+  private final Client              client;
+
+  private final String       workerId;
+  private final SecureRandom random;
+
+  private final UnlockOperation unlockOperation;
+
+  private boolean running;
+  private boolean finished;
+
+
+  public DirectoryReconciler(ContactDiscoveryConfiguration cdsConfig,
```

Comment:
Ideally there would be a test for this, and the test would probably be easier to write if you weren't passing in the configuration object directly.